connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java

```diff
@@ -102,6 +99,71 @@ public static boolean isJar(Path path) {
         return locations;
     }
 
+    public static List<Path> pluginUrls(Path topPath) throws IOException {
+        boolean containsClassFiles = false;
+        Set<Path> archives = new HashSet<>();
+        LinkedList<DirectoryEntry> dfs = new LinkedList<>();
+        Set<Path> visited = new HashSet<>();
+
+        if (isArchive(topPath)) {
+            return Collections.singletonList(topPath);
+        }
+
+        DirectoryStream<Path> topListing = Files.newDirectoryStream(
+                topPath,
+                PLUGIN_PATH_FILTER
+        );
+        dfs.push(new DirectoryEntry(topListing));
+        visited.add(topPath);
+        try {
+            while (!dfs.isEmpty()) {
+                Iterator<Path> neighbors = dfs.peek().iterator;
+                if (!neighbors.hasNext()) {
+                    dfs.pop().stream.close();
+                    continue;
+                }
+
+                Path adjacent = neighbors.next();
+                if (Files.isSymbolicLink(adjacent)) {
+                    Path absolute = Files.readSymbolicLink(adjacent).toRealPath();
+                    if (Files.exists(absolute)) {
+                        adjacent = absolute;
+                    } else {
+                        continue;
+                    }
+                }
+
+                if (!visited.contains(adjacent)) {
+                    visited.add(adjacent);
+                    if (isArchive(adjacent)) {
+                        archives.add(adjacent);
+                    } else if (isClassFile(adjacent)) {
+                        containsClassFiles = true;
+                    } else {
+                        DirectoryStream<Path> listing = Files.newDirectoryStream(
+                                adjacent,
+                                PLUGIN_PATH_FILTER
+                        );
+                        dfs.push(new DirectoryEntry(listing));
+                    }
+                }
+            }
+        } finally {
+            while (!dfs.isEmpty()) {
+                dfs.pop().stream.close();
+            }
+        }
+
+        if (containsClassFiles) {
+            if (archives.isEmpty()) {
+                return Collections.singletonList(topPath);
+            }
+            log.warn("Plugin path contains both java archives and class files. Returning only the"
```

Comment:
I think it's possible we could support mixed modes with a bit more logic to track when you hit directories without any class files, but I think anything like this is unlikely to show up in practice anyway, so I'm happy to leave it as a future improvement.